### PR TITLE
Enhance media-controller with behavior from the model

### PR DESCRIPTION
**Pull request by johnBartos**
### This PR will...
- Have the media-controller listen to and forward Provider events (migrated from the Model)
- Have the program-controller act as a single source for media-controller events
- Removes the media-controller from the model
- Remove media-controller references in View and QOE (and replaces them with events from the Model)
- Have the controller listen to the program-controller for Provider events

### Why is this Pull Request needed?
Background loading requires the program-controller to be able to control which events are propagated to the rest of the player. It now accomplishes this through the media-controller, which is solely responsible for receiving and propagating state updates. In this manner we'll be able to hot swap media-controller instances to control which provider influences the player.

### Anything Else?
I still have a bit of cleanup to do. Let's see what bugs automation uncovers.

### Are there any Pull Requests open in other repos which need to be merged with this?
https://github.com/jwplayer/jwplayer-commercial/pull/4332

#### Addresses Issue(s):
JW8-896

### CODE REVIEW

File: src/js/program/media-controller.js [hunk 1]
```diff
@@ -25,6 +33,7 @@ export default class MediaController {
         mediaModelState.duration = duration;
         model.setProvider(provider);
         model.setMediaModel(mediaModel);
+        model.mediaController = this;
```
Comment (robwalch) [inline src/js/program/media-controller.js]: Is this a good idea considering `model.mediaController` is used in other places and expects `mediaController` not to change? I think we should try to only set static properties on `model`. Anything else should be a model attribute.

File: src/js/program/provider-listener.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            case MEDIA_RATE_CHANGE: {
+                const rate = data.playbackRate;
+                // Check if its a generally usable rate.  Shaka changes rate to 0 when pause or buffering.
+                if (rate > 0) {
+                    model.set('playbackRate', rate);
+                }
+            }
+                return;
+            case MEDIA_TYPE:
+                if (mediaModel.get('mediaType') !== data.mediaType) {
```
Comment (egreaves) [inline src/js/program/provider-listener.js]: We can use `MEDIA_TYPE` or `type` in lieu of `'mediaType'` for better minification. There are other types above and below for which we could make this change.

File: src/js/program/provider-listener.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
+import { _isNaN, _isNumber } from 'utils/underscore';
+import { PLAYER_STATE, STATE_IDLE, MEDIA_VOLUME, MEDIA_MUTE,
+    MEDIA_TYPE, PROVIDER_CHANGED, AUDIO_TRACKS, AUDIO_TRACK_CHANGED,
+    MEDIA_RATE_CHANGE, MEDIA_BUFFER, MEDIA_TIME, MEDIA_LEVELS, MEDIA_LEVEL_CHANGED, MEDIA_ERROR,
+    MEDIA_BEFORECOMPLETE, MEDIA_COMPLETE, MEDIA_META, MEDIA_SEEK, MEDIA_SEEKED,
+    NATIVE_FULLSCREEN } from 'events/events';
+
+export default function ProviderEventListener(mediaController) {
```
Comment (robwalch) [inline src/js/program/provider-listener.js]: This is called `ProviderListener` (not `ProviderEventListener`) externally.

**Comment by robwalch** [comment]
@johnBartos @dannyfinks there appears to be a fair number of [test failures](https://jenkins.jwplayer.com/view/JW7/job/jw7-commercial-cucumber/9047/). I'm going to hold off on merging until you get a chance to look at the captions and test failures.

> [:cucumber:](https://jenkins.jwplayer.com/view/JW7/job/jw7-commercial-cucumber/9047/) Automated Tests UNSTABLE

File: src/js/program/media-controller.js [hunk 4]
```diff
@@ -79,30 +87,84 @@ export default class MediaController {
...
+                item,
+                playReason
+            });
+        });
+    }
+
+    _playbackComplete() {
+        const { provider } = this;
+        this.beforeComplete = false;
+        provider.setState(STATE_COMPLETE);
```
Comment (robwalch) [inline src/js/program/media-controller.js]: Where's the attached check? That's what allows postrolls to play without complete firing until after the postroll is done.

**Pull request merged**